### PR TITLE
Enable multiple simulated imu and magnetometers in gazebo

### PR DESCRIPTION
**Pull request by KonradRudin**
## Describe problem solved by this pull request
A simulation should e able to simulate multiple sensors of the same kind. This is implemented for the gps sensors, however, only one imu and magnetometer can be simulated. They are duplicated on the PX4 side.

## Describe your solution
This PR is coupled to the PR in [sitl_gazebo](https://github.com/PX4/PX4-SITL_gazebo/pull/905), where it is now possible to add multiple imu and magnetometer links. On the PX4 side the handling of the hil_sensor message is updated, where the sensor id in the hil_sensor message is used to determine which imu should currently be updated.

## Test data / coverage
Tested with standard_vtol sitl simulation. Needs more testing.

### CODE REVIEW

**Comment by hamishwillee** [comment]
This is cool. When done, can we have docs for how you turn it on in https://docs.px4.io/main/en/simulation/gazebo.html#usage-configuration-options ? Also, is this going into Gazebo ignition as well? I think we'll be moving in that direction.

File: src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp [hunk 1]
```diff
@@ -1424,20 +1438,55 @@ void SimulatorMavlink::check_failure_injections()
...
+						PX4_WARN("CMD_INJECT_FAILURE, mag %d off", i);
+						_mag_blocked[i] = true;
+						_mag_stuck[i] = false;
+					}
+
+				} else if (instance >= 1 && instance <= MAG_COUNT_MAX) {
+					PX4_WARN("CMD_INJECT_FAILURE, mag %d off", instance - 1);
+					_mag_blocked[instance - 1] = true;
+					_mag_stuck[instance - 1] = false;
+				}
```
Comment (tstastny) [inline src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp:1455]: why is `MAG_COUNT_MAX` inclusive in the conditional and `instance - 1` used in the index? are the sensor instances 1-based?

File: src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp [hunk 2]
```diff
@@ -1424,20 +1438,55 @@ void SimulatorMavlink::check_failure_injections()
...
+						PX4_WARN("CMD_INJECT_FAILURE, mag %d stuck", i);
+						_mag_blocked[i] = false;
+						_mag_stuck[i] = true;
+					}
+
+				} else if (instance >= 1 && instance <= MAG_COUNT_MAX) {
+					PX4_WARN("CMD_INJECT_FAILURE, mag %d stuck", instance - 1);
+					_mag_blocked[instance - 1] = false;
+					_mag_stuck[instance - 1] = true;
+				}
```
Comment (tstastny) [inline src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp:1472]: same question

File: src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp [hunk 3]
```diff
@@ -1424,20 +1438,55 @@ void SimulatorMavlink::check_failure_injections()
...
+						PX4_WARN("CMD_INJECT_FAILURE, mag %d ok", i);
+						_mag_blocked[i] = false;
+						_mag_stuck[i] = false;
+					}
+
+				} else if (instance >= 1 && instance <= MAG_COUNT_MAX) {
+					PX4_WARN("CMD_INJECT_FAILURE, mag %d ok", instance - 1);
+					_mag_blocked[instance - 1] = false;
+					_mag_stuck[instance - 1] = false;
+				}
```
Comment (tstastny) [inline src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp:1489]: same question

**Review by tstastny — CHANGES_REQUESTED** [review]
probably just my misunderstanding on the couple of comments - once addressed or dismissed, looks ok to me!

@dagar i guess once https://github.com/PX4/PX4-Autopilot/pull/20108 goes in, a rebase will handle the flash issue?

File: src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp (hunk 1 above)
Comment (KonradRudin) [reply src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp:1455]: This is analogously implemented as for the gyros. Here you can set failures of the magnetometers and select the instance which should fails. If you select instance 0 this is equivalent as to set the failure on all magnetometers. with Instances 1 to MAG_COUNT_MAX you can set only the specific instance. This is only for the interface to set a failure.

File: src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp (hunk 2 above)
Comment (KonradRudin) [reply src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp:1472]: Same reply :)

File: src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp (hunk 3 above)
Comment (KonradRudin) [reply src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp:1489]: same Reply :)

**Review by tstastny — APPROVED** [review]

**Comment by KonradRudin** [comment]
@tstastny the gazebo side was merged into main. i just updated the sitl_gazebo submodule to point to nexest main including the gazebo multi IMU changes. Secondly i rebased the branch on current main.

**Review by tstastny — APPROVED** [review]

**Pull request merged**